src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -1432,41 +1464,177 @@ int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl
+        return -EINVAL;
+      }
+      off_t lst = std::stoi(size);
+      off_t fst = 0;
+
+      do {
+	std::string prefix = source->get_bucket()->get_name() + "_" + version + "_" + source->get_name();
+	if (fst >= lst) {
+	  break;
+	}
+
+	off_t cur_size = std::min<off_t>(fst + dpp->get_cct()->_conf->rgw_max_chunk_size, lst);
+	off_t cur_len = cur_size - fst;
+	block.blockID = static_cast<uint64_t>(fst);
+	block.size = static_cast<uint64_t>(cur_len);
+
+	if ((ret = blockDir->get(dpp, &block, y)) < 0) {
+	  ldpp_dout(dpp, 0) << "Failed to retrieve directory entry for: " << source->get_name() << " blockid: " << fst << " block size: " << cur_len << ", ret=" << ret << dendl;
+	  return ret;
+	}
+
+	if (block.cacheObj.dirty)
+	  prefix = "D_" + prefix;
 
+	std::string oid_in_cache = prefix + "_" + std::to_string(fst) + "_" + std::to_string(cur_len);
+
+	if ((ret = blockDir->del(dpp, &block, y)) == 0) { 
+	  if ((ret = source->driver->get_cache_driver()->delete_data(dpp, oid_in_cache, y)) == 0) { // Sam: do we want del or delete_data here? 
+	    if (!(ret = source->driver->get_policy_driver()->get_cache_policy()->erase(dpp, policy_prefix + "_" + std::to_string(fst) + "_" + std::to_string(cur_len), y))) {
+	      ldpp_dout(dpp, 0) << "Failed to delete policy entry for: " << source->get_name() << " blockID: " << fst << " block size: " << cur_len << ", ret=" << ret << dendl;
+	      return ret;
+	    }
+	  } else {
+	    ldpp_dout(dpp, 0) << "Failed to delete existing block for: " << source->get_name() << " blockID: " << fst << " block size: " << cur_len << ", ret=" << ret << dendl;
+	    return ret;
+	  }
+	} else if (ret == -ENOENT) {
+	  continue;
+	} else {
+	  ldpp_dout(dpp, 0) << "Failed to delete directory entry for: " << source->get_name() << " blockid: " << fst << " block size: " << cur_len << ", ret=" << ret << dendl;
+	  return ret;
+	}
+
+	fst += cur_len;
+      } while (fst < lst);
+
+      if (!objDirty) { // object written to backend  
+	return next->delete_obj(dpp, y, flags);
+      }
+    } else { // provided version is a delete marker; remove delete marker block
+      rgw::d4n::CacheBlock deleteBlock;
+      deleteBlock = block;
+      block.cacheObj.objName = source->get_name();
+
+      if (block.prevVersion) { // move previous version to current version and update for head object
+        block.version = block.prevVersion->first;
+        block.deleteMarker = block.prevVersion->second;
+        block.prevVersion = {};
+      }
+
+      if ((ret = blockDir->del(dpp, &deleteBlock, y)) == 0) {
+	if ((ret = blockDir->set(dpp, &block, y)) < 0) {
+	  ldpp_dout(dpp, 0) << "Failed to set head object in block directory for: " << block.cacheObj.objName << ", ret=" << ret << dendl;
+	  return ret;
+        }
+      } else {
+	ldpp_dout(dpp, 0) << "Failed to delete delete marker block in block directory for: " << block.cacheObj.objName << ", ret=" << ret << dendl;
+	return ret;
+      }
+    }
```

Comment:
You need to call eraseObj to erase the object from the map/object heap (this only applies to dirty objects).

Reply:
`eraseObj` is called in the `cleaning` method; can you explain why it's called there rather than being updated to not have a dirty flag? The `o_entries_map` would no longer contain this object entry and attempting to delete it during a `delete_obj` call would result in a failure.

Reply:
like I said in my previous comment o_entries_map is used to maintain only dirty objects, hence cleaning calls eraseObj(); there is a call to objDir->update_field, to update the dirty flag correctly. In case delete obj is called for a dirty object, eraseObj() needs to be called for the reason I explained. If it is dirty, erasing it from o_entries_map will not result in a failure.